### PR TITLE
281: Caching for read heavy endpoints

### PR DESCRIPTION
**Pull request by prince-b-b**
## Description

Adds HTTP response caching to all read-heavy API endpoints via a shared  
`cache_read_actions` decorator applied to `list` and `retrieve` actions.

Cache entries vary by `X-Api-Key` and `Accept-Language` header to prevent  
cross-user data leakage and translation mix-ups.

Invalidation is handled centrally via Django signals that clear the cache on  
`post_save` and `post_delete` for all affected models.

## Related Issues

- Issue #281 

## Type of Change

- [X] Feature  
- [X] Internal improvement (cache invalidation reliability)

### Default test suite (no Redis required)

Running `pytest` executes **all tests using LocMemCache** (in-memory cache).  
This is the default and does **NOT require Redis**.

```bash
pytest
```

Or only the cache tests:

```bash
pytest tests/core/test_core_cache.py -v
```

This verifies:
- caching works on list/retrieve endpoints
- cache varies correctly on headers
- write operations bypass cache
- cache invalidation works correctly

---

### Redis integration tests (requires Redis)

⚠️ Important: `pytest -m redis` **alone is NOT sufficient**.  
You must also switch the Django settings to use Redis.

#### 1. Start Redis locally

Using Docker:

```bash
docker run --name test-redis -p 6379:6379 redis:7
```

Or via system service:

```bash
sudo service redis-server start
```

Verify Redis is running:

```bash
redis-cli ping
```

Expected output:

```
PONG
```

---

#### 2. Run Redis-specific tests (with Redis backend)

```bash
DJANGO_SETTINGS_MODULE=config.settings.test_redis pytest -m redis
```

These tests verify:
- cache works with Redis backend
- cached responses are returned correctly
- invalidation works end-to-end with Redis

---

### ⚠️ Common pitfall

Running:

```bash
pytest -m redis
```

without setting `DJANGO_SETTINGS_MODULE=config.settings.test_redis` will still use  
**LocMemCache**, meaning Redis is NOT actually being tested.

---

## Docker-related notes (specifically for @JDLMRN)

For Docker-based environments:

### Minimal Redis service

```yaml
services:
  redis:
    image: redis:7
    ports:
      - "6379:6379"
```

### Environment requirements

The backend expects:

```env
REDIS_URL=redis://redis:6379/1
```

(or `127.0.0.1` if running locally outside Docker)

---

### Running tests inside Docker

Example:

```bash
docker compose up -d redis
docker compose exec backend pytest -m redis --ds=config.settings.test_redis
```

---

### CI recommendation

Split tests into two jobs:

1. **Unit tests (fast, no Redis)**
   ```bash
   pytest -m "not redis"
   ```

2. **Redis integration tests**
   - Requires Redis service
   ```bash
   pytest -m redis --ds=config.settings.test_redis
   ```

---

## Checklist for Reviewers

- [X] Follows Style Guide  
- [X] Has Documentation  
- [X] Added Unit Tests  
- [ ] Tested with LocMemCache (default test suite)  
- [ ] Tested with Redis (integration tests)

### CODE REVIEW

File: backend/config/settings/base.py [hunk 1]
```diff
@@ -339,3 +339,17 @@
...
 }
+
+# ---------------------------------------------------------------------------
+# Caches
+# ---------------------------------------------------------------------------
+
+CACHES = {
+    "default": {
+        "BACKEND": "django_redis.cache.RedisCache",
+        "LOCATION": "redis://127.0.0.1:6379/1",
```
Comment (eliaryck) [inline backend/config/settings/base.py]: You also need to **import os**. It's better because your version was hardcoded.
```suggestion
        "LOCATION": os.getenv("REDIS_URL"),
```

File: backend/apps/core/signals.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        sender=model,
+        weak=False,
+        dispatch_uid=f"{uid_base}:post_save",
+    )
+    post_delete.connect(
+        handler,
+        sender=model,
+        weak=False,
+        dispatch_uid=f"{uid_base}:post_delete",
+    )
```
Comment (eliaryck) [inline backend/apps/core/signals.py:47]: I've now modified the entire file, but actually it only affects the comments and doesn't clear the entire cache; instead, it's more efficient when using a prefix. This way, for example, deleting an event doesn't reset the location in the cache.
```suggestion
"""
Cache invalidation signals for read-heavy API endpoints.

Provides a single helper, ``connect_cache_invalidation``, that wires
post_save and post_delete signals for a given model to a handler that
clears cached responses whenever the underlying data changes.

When using django-redis, only keys matching the url_prefix are deleted.
Falls back to cache.clear() for other backends (e.g. LocMemCache in tests).
"""
from django.core.cache import cache
from django.db.models.signals import post_delete, post_save


def connect_cache_invalidation(model, url_prefix):
    """
    Connect post_save and post_delete signals for the given model to a
    handler that invalidates cached responses for the given url_prefix.

    Uses delete_pattern (django-redis) when available for targeted
    invalidation. Falls back to cache.clear() for other backends.

    Uses stable dispatch_uids so repeated calls do not register duplicate
    receivers during tests or app startup.
    """
    def handler(sender, **kwargs):
        if hasattr(cache, "delete_pattern"):
            cache.delete_pattern(f"*{url_prefix}*")
        else:
            cache.clear()

    model_label = model._meta.label_lower
    uid_base = f"cache_invalidation:{model_label}:{url_prefix}"

    post_save.connect(
        handler,
        sender=model,
        weak=False,
        dispatch_uid=f"{uid_base}:post_save",
    )
    post_delete.connect(
        handler,
        sender=model,
        weak=False,
        dispatch_uid=f"{uid_base}:post_delete",
    )
```

File: backend/apps/core/cache_mixin.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
+"""Provides a mixin to apply HTTP response caching to list and retrieve actions of Django REST Framework viewsets."""
+
+from django.utils.decorators import method_decorator
+from django.views.decorators.cache import cache_page
+from django.views.decorators.vary import vary_on_headers
+
+CACHE_TTL = 60 * 5  # 5 minutes (how long to cache responses for)
```
Comment (eliaryck) [inline backend/apps/core/cache_mixin.py]: You also need to add it in the settings files.
```suggestion
CACHE_TTL = getattr(settings, "CACHE_TTL", 60 * 5)  # default: 5 minutes
```

File: backend/apps/core/cache_mixin.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+def cache_read_actions(ttl=CACHE_TTL):
+    """
+    Class decorator that applies HTTP response caching to list and retrieve.
+    Cache varies on Authorization and Accept-Language so users never receive
+    each other's data and translations stay correct.
+    """
+    decorators = [
+        method_decorator(cache_page(ttl)),
+        method_decorator(vary_on_headers("Authorization", "Accept-Language")),
```
Comment (eliaryck) [inline backend/apps/core/cache_mixin.py]: We use the X-Api-Key header :)
```suggestion
        method_decorator(vary_on_headers("X-Api-Key", "Accept-Language")),
```

File: backend/config/settings/test_redis.py [hunk 5]
```diff
@@ -0,0 +1,13 @@
+"""Gives you a separate Redis-backed test mode for integration checks."""
+
+from .test import *  # noqa: F401, F403
+
+CACHES = {
+    "default": {
+        "BACKEND": "django_redis.cache.RedisCache",
+        "LOCATION": "redis://127.0.0.1:6379/1",
```
Comment (eliaryck) [inline backend/config/settings/test_redis.py]: You also need to import os.
```suggestion
        "LOCATION": os.getenv("REDIS_URL", default="redis://127.0.0.1:6379/1"),
```

File: backend/requirements/base.txt [hunk 6]
```diff
@@ -10,3 +10,4 @@ bleach==6.3.0
 tqdm==4.67.3
 Pillow==12.1.1
 django-cors-headers==4.9.0
+django-redis>=5.4.0
```
Comment (eliaryck) [inline backend/requirements/base.txt]: Use the most recent version. Dependabot will update it later.

File: backend/tests/core/test_core_cache.py [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+        codes = [item["code"] for item in results]
+
+        self.assertIn("de", codes)
+
+
+class TestCacheInvalidationHandler(TestCase):
+    """Tests that the signal handler clears the cache after model changes."""
+
+    def test_post_save_clears_cache(self):
+        mock_cache = MagicMock(spec=["clear"])
```
Comment (eliaryck) [inline backend/tests/core/test_core_cache.py:191]:
```suggestion
        mock_cache = MagicMock(spec=["delete_pattern"])
```

File: backend/tests/core/test_core_cache.py [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+
+        with patch("apps.core.signals.cache", mock_cache):
+            connect_cache_invalidation(Language, "/api/v1/languages/")
+            post_save.send(sender=Language, instance=MagicMock(), created=True)
+
+        mock_cache.clear.assert_called_once()
+
+    def test_post_delete_clears_cache(self):
+        mock_cache = MagicMock(spec=["clear"])
```
Comment (eliaryck) [inline backend/tests/core/test_core_cache.py:200]:
```suggestion
        mock_cache = MagicMock(spec=["delete_pattern"])
```

File: backend/tests/core/test_core_cache.py [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+
+        with patch("apps.core.signals.cache", mock_cache):
+            connect_cache_invalidation(Language, "/api/v1/languages/")
+            from django.db.models.signals import post_delete
```
Comment (eliaryck) [inline backend/tests/core/test_core_cache.py]: I think it's not needed.
```suggestion

```

**Review by eliaryck — CHANGES_REQUESTED** [review]
Just a few small changes :)

**Comment by prince-b-b** [comment]
## Update: changes after review / debugging

After additional testing (especially with Redis), several important improvements were made to the caching implementation:

### 1. Reliable cache invalidation (main fix)

**Before:**
- invalidation was based on URL patterns (`"/api/v1/..."`)
- this was **not reliable with Redis**, because Django builds cache keys internally

**Now:**
- caching uses **explicit cache prefixes** (e.g. `api:languages`)
- invalidation is done via:

```python
cache.delete_pattern(f"{prefix}:*")
```

- this ensures that **only the correct cache entries are removed**

👉 Result: Redis tests no longer fail and invalidation works correctly.

---

### 2. Custom caching instead of `cache_page`

**Before:**
- used `cache_page(...)`
- no control over cache keys → difficult to invalidate

**Now:**
- custom caching logic in `cache_read_actions`
- cache keys include:
  - endpoint prefix
  - query params
  - `X-Api-Key`
  - `Accept-Language`

👉 Result:
- full control over caching
- safe variation per user/language
- compatible with targeted invalidation

---

### 3. Improved signals

- `connect_cache_invalidation` now uses:
  - `delete_pattern` (Redis)
  - fallback to `cache.clear()` (other backends)
- uses **stable `dispatch_uid`s** (prevents duplicate signal handlers in tests)

---

### 4. Tests fixed and extended

**Issues resolved:**
- tests always expected `cache.clear()` → incorrect with Redis
- missing `post_delete` import
- mocks did not include all required methods

**Now:**
- separate tests for:
  - Redis (`delete_pattern`)
  - fallback (`clear`)
- Redis integration test works end-to-end

---

### Small breaking change

If you were manually using cache invalidation:

```python
connect_cache_invalidation(Language, "/api/v1/languages/")
```

it must now be:

```python
connect_cache_invalidation(Language, "api:languages")
```

---

### Why this change?

- `delete_pattern("*url*")` was unreliable
- `cache.clear()` was correct but inefficient
- new approach = **correct + performant + scalable**

---

### Status

- ✔️ Works with LocMemCache
- ✔️ Works with Redis
- ✔️ All tests pass
- ✔️ Invalidation works correctly in all cases

---

If anyone sees edge cases or has suggestions for further improvements, feel free to share.
